### PR TITLE
[Program Migration] Add custom body parser and trim json before submission

### LINKED ISSUE #9249: [Program Migration] Allow importing very large programs
Arkansas ran into an issue importing a very large program. We need to up the limit on request sizes we will accept for program import and/or shrink the request sizes by trimming the json.

### PR DESCRIPTION
**Pull request by rockycodes**
### Description

CEs are still running into issues when importing very large programs. This PR addresses that by adding a custom body parser that will allow larger requests, and by shrinking the size of the request by trimming extra whitespace off the json before submission.

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)

- [x] Added the correct label: < feature | enhancement | bug | under-development | dependencies | infrastructure | ignore-for-release | database >
- [x] Assigned to a specific person, `civiform/developers`, or a [more specific round-robin list](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#adding-reviewers)
- [x] Added an additional reviewer from outside your organization as FYI (if the primary reviewer is in the same organization as you)
- [x] Removed the release notes section if the title is sufficient for the release notes description, or put more details in that section.
- [x] Created unit and/or browser tests which fail without the change (if possible)
- [x] Performed manual testing (Chrome and Firefox if it includes front-end changes)
- [x] Extended the README / documentation, if necessary. For user-facing features, consider updating [the user docs](https://github.com/civiform/docs). For "under-the-hood" changes or things more relevant to developers, consider updating [the dev wiki](https://github.com/civiform/civiform/wiki).
- [x] Ensured PII wasn't added to any new logs, unless it was guarded by `isDevOrStaging`

### Instructions for manual testing

Ask me for large json if you need some to test with!

### Issue(s) this completes

Fixes #9249

### CODE REVIEW

File: server/app/assets/javascripts/admin_import_view.ts [hunk 1]
```diff
@@ -3,13 +3,30 @@ import {attachRedirectToPageListeners} from './main'
...
+    this.addStringifyJsonListener()
   }
 
   addRedirectListeners() {
     document.addEventListener('htmx:afterSwap', () => {
       attachRedirectToPageListeners()
     })
   }
+
+  addStringifyJsonListener() {
```
Comment (rockycodes) [inline server/app/assets/javascripts/admin_import_view.ts:15]: I couldn't think of a way to test the changes I'm making here. Lmk if you have ideas!

File: server/app/parsers/ProgramMigrationParser.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import play.mvc.BodyParser;
+
+/**
+ * This is a custom HTTP request body parser for Program Migration that increases the allowable
+ * limit for buffering. Program json is often quite large and we want to allow it to be processed
+ * without increasing the buffer limit for all HTTP requests. All other requests use the default
+ * Play buffer limit which is 100KB. For more information see:
+ * https://www.playframework.com/documentation/3.0.x/JavaBodyParsers
+ */
+public class ProgramMigrationParser extends BodyParser.FormUrlEncoded {
```
Comment (gwendolyngoetz) [inline server/app/parsers/ProgramMigrationParser.java:14]: (blocking) There's nothing specific about this for program migration. Can you:
- move the program migration specific comments to the `hxImportProgram` controller action 
- rename this to something like `LargeBodyParser` or `LargeFormUrlEncodedBodyParser`
- include in the javadoc that this supports 1 MiB (mebibyte)

File: server/app/views/admin/migration/AdminImportView.java [hunk 3]
```diff
@@ -32,14 +32,15 @@
  */
 public class AdminImportView extends BaseHtmlView {
   /**
-   * Play Framework defaults to 100KB memory limit per request. We set this to 256KB in
-   * application.conf to allow large programs to be imported. More info:
+   * Play Framework defaults to 100KB memory limit per request which is too small for many programs.
+   * We set this to 512KB in {@link parsers.ProgramMigrationParser} to allow large programs to be
```
Comment (gwendolyngoetz) [inline server/app/views/admin/migration/AdminImportView.java:36]:
```suggestion
   * We use a custom body parser {@link parsers.ProgramMigrationParser} to allow large programs to be
```
The size in KB differs from what the parser is configured for. I'd say leave the actual value out of this comment so it'll never get out of sync.

**Review by gwendolyngoetz — CHANGES_REQUESTED** [review]

File: server/app/views/admin/migration/AdminImportView.java [hunk 4]
```diff
@@ -32,14 +32,15 @@
...
+   * imported. More info:
    * https://www.playframework.com/documentation/2.9.x/JavaBodyParsers#Content-length-limits
    *
    * <p>Pasting in a json string above this limit will automatically truncate the string which
    * results in a server error.
    */
-  private static final int MAX_TEXT_LENGTH = 256000;
+  private static final int MAX_TEXT_LENGTH = 512000;
```
[the remark this answers is not in the document]
Comment (rockycodes) [reply server/app/views/admin/migration/AdminImportView.java:43]: Good idea. Updated!

File: server/app/assets/javascripts/admin_import_view.ts [hunk 5]
```diff
@@ -3,13 +3,30 @@ import {attachRedirectToPageListeners} from './main'
...
+      const trimmedProgramJson = JSON.stringify(JSON.parse(programJson))
+      customEvent.detail.formData.set('programJson', trimmedProgramJson)
+    })
+  }
+}
+
+// htmx does not provide Typescript definitions for its custom events, so we have to define them ourselves
+// For more info about fields on an htmx request see https://htmx.org/events/#htmx:configRequest
+interface HtmxRequest {
+  formData: FormData
```
[the remark this answers is not in the document]
Comment (rockycodes) [reply server/app/assets/javascripts/admin_import_view.ts]: done!

File: server/app/parsers/ProgramMigrationParser.java (hunk 2 above)
Comment (rockycodes) [reply server/app/parsers/ProgramMigrationParser.java]: done!

**Pull request merged**